ajax/asset/assetdefinition.php

```diff
@@ -0,0 +1,104 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+use Glpi\Asset\AssetDefinition;
+use Glpi\Asset\CustomFieldDefinition;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/** @var \Glpi\Controller\LegacyFileLoadController $this */
+$this->setAjax();
+
+Session::checkRight(AssetDefinition::$rightname, READ);
+
+if ($_GET['action'] === 'get_all_fields') {
+    header("Content-Type: application/json; charset=UTF-8");
+    $definition = new AssetDefinition();
+    if (!$definition->getFromDB($_GET['assetdefinitions_id'])) {
+        http_response_code(404);
+        exit();
+    }
+    $all_fields = $definition->getAllFields();
+    $field_results = [];
+    foreach ($all_fields as $k => $v) {
+        $field_info = is_array($v) ? $v : ['text' => $v];
+        if (!empty($_POST['searchText']) && stripos($field_info['text'], $_POST['searchText']) === false) {
+            continue;
+        }
+        $field_info['id'] = $k;
+        $field_results[] = $field_info;
+    }
+    echo json_encode([
+        'results' => $field_results,
+        'count' => count($all_fields)
+    ], JSON_THROW_ON_ERROR);
+    exit();
+} else if ($_GET['action'] === 'get_field_placeholder') {
+    header("Content-Type: text/html; charset=UTF-8");
+    $custom_field = new CustomFieldDefinition();
+    if ($_GET['customfields_id'] > 0) {
+        if (!$custom_field->getFromDB($_GET['customfields_id'])) {
+            throw new NotFoundHttpException();
+        }
+    } else {
+        $custom_field->fields['name'] = '';
+        $custom_field->fields['label'] = $_GET['label'];
+        $custom_field->fields['type'] = $_GET['type'];
+        $custom_field->fields['itemtype'] = 'Computer'; // Doesn't matter what it is as long as it's not empty
+        $custom_field->fields['default_value'] = '';
+
+        $asset_definition = new AssetDefinition();
+        if (!$asset_definition->getFromDB($_GET['assetdefinitions_id'])) {
+            throw new NotFoundHttpException();
+        }
+        $fields_display = $asset_definition->getDecodedFieldsField();
+        foreach ($fields_display as $field) {
+            if ($field['key'] === $_GET['key']) {
+                $custom_field->fields['field_options'] = $field['field_options'] ?? [];
+                break;
+            }
+        }
+    }
+    $custom_field->fields['field_options']['disabled'] = true;
+    echo $custom_field->getFieldType()->getFormInput('', null);
+    exit();
+} else if ($_GET['action'] === 'get_core_field_editor') {
+    header("Content-Type: text/html; charset=UTF-8");
+    $asset_definition = new AssetDefinition();
+    if (!$asset_definition->getFromDB($_GET['assetdefinitions_id'])) {
+        throw new NotFoundHttpException();
+    }
+    $asset_definition->showFieldOptionsForCoreField($_GET['key']);
+    exit();
+}
+http_response_code(400);
```

Comment:
If we remove the `exit` statement, this need to be in an `else`.

```suggestion
} else {
    throw new BadRequestHttpException();
}
```

Reply:
Cautious about throwing exceptions in AJAX files. I already had to revert some not found exceptions to 404 response code + exit because GLPI is unnecessarily sending an entire error page as a response body.

Functionally, maybe not an issue. Performance wise, I imagine it isn't great to be doing this.